### PR TITLE
add test script for testing pa functionality

### PR DESCRIPTION
**Pull request by biox**
- it's a shell script - run it using ./test
- it creates a dir `test-stuff` that contains the test artifacts
- the dir can be inspected between runs to help with debugging

### CODE REVIEW

File: test [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+# clean up previous run
+# (the previous state is left around
+# intentionally, in case the dev
+# wants to poke around)
+# that's why we don't clean up on exit
+rm -rf test-stuff
+
+# pa welcomes you
+$pa | grep -q "a simple password manager" ||
```
[the remark this answers is not in the document]
Comment (biox) [reply test:17]: i used an env var because some tests require `cd`ing, which would mean using `../../pa` - do you think that's preferable?

File: test [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+test -s test-stuff/identities ||
+    printf "an identities file should exist\n"
+test -s test-stuff/recipients ||
+    printf "a recipients file should exist\n"
+test -d "$PA_DIR/.git" ||
+    printf "git dir should exist\n"
+# TODO: ensure git author/email are set correctly, etc
+
+# pa add
+printf 'y\n' | $pa add test 2>&1 >/dev/null ||
```
Comment (arcxio) [inline test:33]: \n is not necessary, thanks to `-icanon`

File: test [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+# pa welcomes you
+$pa | grep -q "a simple password manager" ||
+    printf "pa should print a welcome message\n"
+
+# generate pa dirs/identityfile/recipientfile
+$pa list
+
+# pa auto-generated files are correct
+test -s test-stuff/identities ||
```
[the remark this answers is not in the document]
Comment (biox) [reply test:24]: meant to be TDD in action lmao, i figure i might have time for the `PA_DIR` PR pretty soon, so maybe i figure this will remind and encourage me to do it haha.

File: test [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+test -d "$PA_DIR/.git" ||
+    printf "git dir should exist\n"
+# TODO: ensure git author/email are set correctly, etc
+
+# pa add
+printf 'y\n' | $pa add test 2>&1 >/dev/null ||
+    printf "pa add should be capable of adding a test password\n"
+printf 'y\n' | $pa add nested/password 2>&1 |
+    grep -q "Saved 'nested/password' to the store." ||
+    printf 'pa add should say it stored nested/passworod\n'
```
[the remark this answers is not in the document]
Comment (biox) [reply test]: fixed in [b5a506b](https://github.com/biox/pa/pull/48/commits/b5a506b350d7d94501fd63432b69fac97ac8d8be)

File: test (hunk 1 above)
[the remark this answers is not in the document]
Comment (arcxio) [reply test:17]: I see that the only one requiring cd is a git one, I think `git -C test-stuff/passwords log` is a better way there

File: test (hunk 3 above)
[the remark this answers is not in the document]
Comment (arcxio) [reply test:24]: fair enough

File: test (hunk 2 above)
Comment (biox) [reply test:33]: fixed in df212eb

Comment (biox) [reply test]: fixed in [df212eb](https://github.com/biox/pa/pull/48/commits/df212eb779a0bf031940883fa5cb8f4a254798df)

File: test (hunk 1 above)
[the remark this answers is not in the document]
Comment (biox) [reply test]: agreed [f168274](https://github.com/biox/pa/pull/48/commits/f168274a64d66763a018673a74198d7872e27906)

File: test [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    printf "a recipients file should exist\n"
+test -d "$PA_DIR/.git" ||
+    printf "git dir should exist\n"
+# TODO: ensure git author/email are set correctly, etc
+
+# pa add
+printf 'y\n' | $pa add test 2>&1 >/dev/null ||
+    printf "pa add should be capable of adding a test password\n"
+printf 'y\n' | $pa add nested/password 2>&1 |
+    grep -q "Saved 'nested/password' to the store." ||
```
[the remark this answers is not in the document]
Comment (biox) [reply test:34]: not sure what you mean here - storing a var and doing a compare seems a lot harder than just grepping to me?

[the remark this answers is not in the document]
Comment (arcxio) [reply test:34]: I mean like this
```shell
test "$(printf y | $pa add nested/password 2>&1)" = "\
Generate a password? [y/N]: y
Saved 'nested/password' to the store." ||
```
it is harder but it's more robust to check for the whole output

[the remark this answers is not in the document]
Comment (biox) [reply test]: bleh, i'm not opinionated either way. i implemented it for a few that i think it could be useful for in 63cc519

**Comment by arcxio** [comment]
i like this, more tests are needed but this is a good start

**Comment by Sjlver** [comment]
This is nice! I love how simple it is.

My suggestion would be to add something like this:

1. Create an error function like this

```shell
failures=0

fail() {
  printf "$@" >&2
  failures=$((failures + 1))
}
```

2. Add this at the end:

```
# Return with a success status if all tests pass
test "$failures" -eq 0
```

Having a "correct" exit status makes it possible to use various integrations, or run the tests as part of a git hook, etc.

You could also add "set -e" at the start, to also fail on unexpected errors.

Again: great work!

**Comment by biox** [comment]
yep good call, the `fail` func makes the code more readable too - added in [2c9441e](https://github.com/biox/pa/pull/48/commits/2c9441e3ed015b27216df867fc5583a7140c0303)

File: test [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+# clean up previous run
+# (the previous state is left around
+# intentionally, in case the dev
+# wants to poke around)
+# that's why we don't clean up on exit
+rm -rf test-stuff
+
+# pa welcomes you
+./pa | grep -q "a simple password manager" ||
+    fail "pa should print a welcome message\n"
```
Comment (arcxio) [inline test:22]: redundant \n

Comment (biox) [reply test]: good catch. fixed in [06b996a](https://github.com/biox/pa/pull/48/commits/06b996a4d8deb35e90e44afa9b322e5180f2d021)

File: test [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+test -d "$PA_DIR/.git" ||
+    fail "git dir should exist"
+# TODO: ensure git author/email are set correctly, etc
+
+# pa add
+printf 'y' | ./pa add test 2>&1 >/dev/null ||
+    fail "pa add should be capable of adding a test password"
+
+test "$(printf y | ./pa add nested/password 2>&1)" = "\
+Generate a password? [y/N]: y
```
Comment (arcxio) [inline test:43]: this output is supposed to be lowercased now

Comment (biox) [reply test]: corrected!
